database/postgres.go

```diff
@@ -0,0 +1,145 @@
+package database
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"net"
+
+	"cloud.google.com/go/alloydbconn"
+	"github.com/jackc/pgx/v5"
+	"github.com/jackc/pgx/v5/pgconn"
+	"github.com/jackc/pgx/v5/stdlib"
+	_ "github.com/jackc/pgx/v5/stdlib"
+	"github.com/moov-io/base/log"
+)
+
+const (
+	// PostgreSQL Error Codes
+	// https://www.postgresql.org/docs/current/errcodes-appendix.html
+	postgresErrUniqueViolation = "23505"
+)
+
+func postgresConnection(ctx context.Context, logger log.Logger, config PostgresConfig, databaseName string) (*sql.DB, error) {
+	var connStr string
+	if config.Alloy != nil {
+		c, err := getAlloyDBConnectorConnStr(ctx, config, databaseName)
+		if err != nil {
+			return nil, logger.LogErrorf("creating alloydb connection: %w", err).Err()
+		}
+		connStr = c
+	} else {
+		c, err := getPostgresConnStr(config, databaseName)
+		if err != nil {
+			return nil, logger.LogErrorf("creating postgres connection: %w", err).Err()
+		}
+		connStr = c
+	}
+
+	db, err := sql.Open("pgx", connStr)
+	if err != nil {
+		return nil, logger.LogErrorf("opening database: %w", err).Err()
+	}
+
+	err = db.Ping()
+	if err != nil {
+		_ = db.Close()
+		return nil, logger.LogErrorf("connecting to database: %w", err).Err()
+	}
+
+	return db, nil
+}
+
+func getPostgresConnStr(config PostgresConfig, databaseName string) (string, error) {
+	url := fmt.Sprintf("postgres://%s:%s@%s/%s", config.User, config.Password, config.Address, databaseName)
+
+	params := ""
+
+	if config.TLS != nil {
+		params += "sslmode=verify-full"
+
+		if config.TLS.CACertFile == "" {
+			return "", fmt.Errorf("missing TLS CA file")
+		}
+		params += "&sslrootcert=" + config.TLS.CACertFile
+
+		if config.TLS.ClientCertFile != "" {
+			params += "&sslcert=" + config.TLS.ClientCertFile
+		}
+
+		if config.TLS.ClientKeyFile != "" {
+			params += "&sslkey=" + config.TLS.ClientKeyFile
+		}
+	}
+
+	connStr := fmt.Sprintf("%s?%s", url, params)
+	return connStr, nil
+}
+
+func getAlloyDBConnectorConnStr(ctx context.Context, config PostgresConfig, databaseName string) (string, error) {
+	if config.Alloy == nil {
+		return "", fmt.Errorf("missing alloy config")
+	}
+
+	var dialer *alloydbconn.Dialer
+	var dsn string
+
+	if config.Alloy.UseIAM {
+		d, err := alloydbconn.NewDialer(ctx, alloydbconn.WithIAMAuthN())
+		if err != nil {
+			return "", fmt.Errorf("creating alloydb dialer: %v", err)
+		}
+		dialer = d
+		dsn = fmt.Sprintf(
+			// sslmode is disabled because the alloy db connection dialer will handle it
+			// no password is used with IAM
+			"user=%s dbname=%s sslmode=disable",
```

Comment:
Is there a reason we format a DSN here but for vanilla postgres use their URI syntax?

Reply:
Either should work. I just preferred the `key=value` syntax when not giving a server (since thats handled by the `alloydbconn` library when connecting to a hosted AlloyDB instance)